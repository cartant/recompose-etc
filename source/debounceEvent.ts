/**
 * @license Use of this source code is governed by an MIT-style license that
 * can be found in the LICENSE file at https://github.com/cartant/recompose-etc
 */

import * as React from "react";
import {
  mapper,
  setDisplayName,
  wrapDisplayName
} from "recompose";
import { debounceTime, distinctUntilChanged, map, tap } from "rxjs/operators";
import { transformEvent } from "./transformEvent";
import { EqualComparer, HandlerProp, RenderProp } from "./types";

export function debounceEvent<TInnerEvent, TOuterEvent>(
  duration: number,
  distinct: boolean | EqualComparer<TOuterEvent>,
  transform: mapper<TInnerEvent, TOuterEvent>
): React.ComponentType<
  HandlerProp<TOuterEvent> &
  RenderProp<HandlerProp<TInnerEvent>>
> & { Props: HandlerProp<TInnerEvent> };

export function debounceEvent<TEvent>(
  duration: number,
  distinct?: boolean | EqualComparer<TEvent>
): React.ComponentType<
  HandlerProp<TEvent> &
  RenderProp<HandlerProp<TEvent>>
> & { Props: HandlerProp<TEvent> };

export function debounceEvent<TInnerEvent, TOuterEvent>(
  duration: number,
  distinct: undefined | boolean | EqualComparer<TOuterEvent>,
  transform?: mapper<TInnerEvent, TOuterEvent>
): React.ComponentType<
  HandlerProp<TOuterEvent> &
  RenderProp<HandlerProp<TInnerEvent>>
> & { Props: HandlerProp<TInnerEvent> } {
  const operators = [
    tap((event: any) => event && event.persist && event.persist()),
    debounceTime(duration)
  ];
  if (transform) {
    operators.push(map(transform));
  }
  if (distinct) {
    const comparer = (typeof distinct === "function") ? distinct : defaultEqualComparer;
    operators.push(distinctUntilChanged(comparer));
  }
  const Component = transformEvent<TInnerEvent, TOuterEvent>(
    event$ => event$.pipe(...operators)
  );
  if (process.env.NODE_ENV !== "production") {
    return setDisplayName<any>(wrapDisplayName(Component, "debounceEvent"))(Component) as any;
  }
  return Component;
}

function defaultEqualComparer(left: any, right: any): boolean {
  if (!left || !left.target || !right || !right.target) {
    return left === right;
  }
  return left.target.value === right.target.value;
}
