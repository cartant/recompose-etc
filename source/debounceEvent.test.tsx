/**
 * @license Use of this source code is governed by an MIT-style license that
 * can be found in the LICENSE file at https://github.com/cartant/recompose-etc
 */

import { shallow } from "enzyme";
import * as React from "react";
import * as renderer from "react-test-renderer";
import { marbles } from "rxjs-marbles";
import { debounceEvent } from "./debounceEvent";

describe("duration only", () => {

  type Event = React.ChangeEvent<HTMLInputElement>;
  const Debounce = debounceEvent<Event>(100);

  it("should debounce events", marbles((m) => {
    m.bind();
    let changeEvents: any[] = [];
    const wrapper = shallow(
      <Debounce
        handler={event => changeEvents.push(event)}
        render={({ handler }) => <input id="input" onChange={handler} type="text" />}
      />
    );
    wrapper.find("#input").simulate("change", { target: { value: "1" } });
    wrapper.find("#input").simulate("change", { target: { value: "12" } });
    wrapper.find("#input").simulate("change", { target: { value: "123" } });
    m.scheduler.schedule(() => {
      wrapper.find("#input").simulate("change", { target: { value: "1234" } });
      wrapper.find("#input").simulate("change", { target: { value: "123" } });
    }, 200);
    m.flush();
    expect(changeEvents).toEqual([
      { target: { value: "123" } },
      { target: { value: "123" } }
    ]);
  }));

  it("should render correctly", () => {
    const rendering = renderer.create(
      <Debounce
        handler={() => {}}
        render={({ handler }) => <input id="input" onChange={handler} type="text" />}
      />
    ).toJSON();
    expect(rendering).toMatchSnapshot();
  });
});

describe("default distinct", () => {

  type Event = React.ChangeEvent<HTMLInputElement>;
  const Debounce = debounceEvent<Event>(100, true);

  it("should debounce events", marbles((m) => {
    m.bind();
    let changeEvents: any[] = [];
    const wrapper = shallow(
      <Debounce
        handler={event => changeEvents.push(event)}
        render={({ handler }) => <input id="input" onChange={handler} type="text" />}
      />
    );
    wrapper.find("#input").simulate("change", { target: { value: "1" } });
    wrapper.find("#input").simulate("change", { target: { value: "12" } });
    wrapper.find("#input").simulate("change", { target: { value: "123" } });
    m.scheduler.schedule(() => {
      wrapper.find("#input").simulate("change", { target: { value: "1234" } });
      wrapper.find("#input").simulate("change", { target: { value: "123" } });
    }, 200);
    m.flush();
    expect(changeEvents).toEqual([{ target: { value: "123" } }]);
  }));
});

describe("distinct using comparer", () => {

  type Event = React.ChangeEvent<HTMLInputElement>;
  const Debounce = debounceEvent<Event>(100, (left, right) => left.target["value"] === right.target["value"]);

  it("should debounce events", marbles((m) => {
    m.bind();
    let changeEvents: any[] = [];
    const wrapper = shallow(
      <Debounce
        handler={event => changeEvents.push(event)}
        render={({ handler }) => <input id="input" onChange={handler} type="text" />}
      />
    );
    wrapper.find("#input").simulate("change", { target: { value: "1" } });
    wrapper.find("#input").simulate("change", { target: { value: "12" } });
    wrapper.find("#input").simulate("change", { target: { value: "123" } });
    m.scheduler.schedule(() => {
      wrapper.find("#input").simulate("change", { target: { value: "1234" } });
      wrapper.find("#input").simulate("change", { target: { value: "123" } });
    }, 200);
    m.flush();
    expect(changeEvents).toEqual([{ target: { value: "123" } }]);
  }));
});

describe("distinct and transformed", () => {

  type Event = React.ChangeEvent<HTMLInputElement>;
  const Debounce = debounceEvent<Event, string>(100, true, event => event.target["value"]);

  it("should debounce events", marbles((m) => {
    m.bind();
    let changeEvents: any[] = [];
    const wrapper = shallow(
      <Debounce
        handler={event => changeEvents.push(event)}
        render={({ handler }) => <input id="input" onChange={handler} type="text" />}
      />
    );
    wrapper.find("#input").simulate("change", { target: { value: "1" } });
    wrapper.find("#input").simulate("change", { target: { value: "12" } });
    wrapper.find("#input").simulate("change", { target: { value: "123" } });
    m.scheduler.schedule(() => {
      wrapper.find("#input").simulate("change", { target: { value: "1234" } });
      wrapper.find("#input").simulate("change", { target: { value: "123" } });
    }, 200);
    m.flush();
    expect(changeEvents).toEqual(["123"]);
  }));
});
