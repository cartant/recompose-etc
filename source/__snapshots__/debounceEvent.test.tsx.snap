// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`duration only should render correctly 1`] = `
<input
  id="input"
  onChange={[Function]}
  type="text"
/>
`;
